### PR TITLE
[10.0][ADD] Add default analytic account for accounts

### PR DESCRIPTION
**Pull request by p-tombez**
Add the ability to set default analytic account on accounts.
This is useful when, e.g. you use the account_analytic_required
and have multi-currencies, thus need to create an automatic
currency exchange difference entry.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ Available addons
 ----------------
 addon | version | summary
 --- | --- | ---
+[account_analytic_default_account](account_analytic_default_account/) | 10.0.1.0.0 | Add default analytic account for accounts
```
Comment (simahawk) [inline README.md]: skip it: this part is updated automatically ;)

File: account_analytic_default_account/README.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Account Analytic Default Account
+================================
+
+This module adds the ability to set a default analytic account on
+accounts.
+
+Configuration
+=============
+
+For example, if you want to have an analytic account whenever you use
```
Comment (simahawk) [inline account_analytic_default_account/README.rst]: skip the 1st part. Just state "Go to Accounting..."

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+from odoo import api, fields, models
+
+
+class AccountAnalyticDefaultAccount(models.Model):
+    _inherit = "account.analytic.default"
+
+    account_id = fields.Many2one(
+        'account.account', string='Account',
+        ondelete='cascade',
+        help="Select an account which will use analytic account specified in "
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: can you rephrase this? The way I get is: "This account will be used by default for sale orders or invoices" or something similar :)

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+             "analytic default (e.g. create new customer invoice or "
+             "Sales order if we select this product, "
+             "it will automatically take this as an analytic account)"
+    )
+
+    @api.model
+    def account_get(self, product_id=None, partner_id=None, user_id=None,
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: I'd split this in more methods and simplify it a bit, for instance:
```python
    account_get_domain_keys = (
        'product_id', 'partner_id', 'user_id',
        'date', 'company_id', 'account_id',
    )

    def _account_get_domain(self, **kw):
        """Build account.analytic.default domain.

        :param kw: filter keys.
            Available filter keys are defined into `account_get_domain_keys`.
        :return: a search domain with all required keys' leaves.
            Each key will be searched for both False and specified value.
            For instance: [
                ('product_id'), '=', False), '|', [('product_id'), '=', 100)
            ]
        """
        domain = []
        account_get_domain_keys = self.account_get_domain_keys[:]
        # date will be handled differently
        date_key = account_get_domain_keys.pop('date', None)

        for key in account_get_domain_keys:
            if kw.get(key):
                domain += ['|', (key, '=', kw.get(key))]
            domain += [(key, '=', False)]

        if date_key and kw.get(date_key):
            domain += ['|', ('date_start', '<=', kw.get(date_key)),
                       ('date_start', '=', False)]
            domain += ['|', ('date_stop', '>=', kw.get(date_key)),
                       ('date_stop', '=', False)]
        return domain

    def account_get(self, **kw):
        domain = self._account_get_domain(**kw)
        best_index = -1
        res = self.env['account.analytic.default']
        for rec in self.search(domain):
            index = 0
            for k in kw.keys():
                if rec[k]:
                    index += 1
            if index > best_index:
                res = rec
                best_index = index
        return res
```

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = "account.invoice.line"
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        res = super(AccountInvoiceLine, self)._onchange_product_id()
+        rec = self.env['account.analytic.default'].account_get(
+            self.product_id.id, self.invoice_id.partner_id.id, self.env.uid,
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: the method is supposed to accept keyword args

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            fields.Date.today(), company_id=self.company_id.id,
+            account_id=self.account_id.id)
+        self.account_analytic_id = rec.analytic_id.id
+        return res
+
+    def _set_additional_fields(self, invoice):
+        if not self.account_analytic_id:
+            rec = self.env['account.analytic.default'].account_get(
+                self.product_id.id, self.invoice_id.partner_id.id,
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: same

**Review by simahawk — CHANGES_REQUESTED** [review]

**Comment by simahawk** [comment]
would you mind adding some tests for this?

File: account_analytic_default_account/models/account_analytic_default_account.py (hunk 4 above)
Comment (p-tombez) [reply account_analytic_default_account/models/account_analytic_default_account.py]: I can't change the method signature so drastically as it already exists in the Odoo addons. I can't guarantee that the calls are made with kwargs everywhere. That's the main reason I copied the method and updated it for my needs.

Comment (simahawk) [reply account_analytic_default_account/models/account_analytic_default_account.py]: fine, you are right. Keep the signature as it is... but respect it in your calls if possible :)
Then you can still address the rest I guess ;)

**Comment by p-tombez** [comment]
@simahawk Thanks for the suggestions! I also added some tests

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        ondelete='cascade',
+        help="Select the account corresponding to an analytic account "
+             "which will be used on the lines of invoices or account moves"
+    )
+
+    def _account_get_domain_keys(self):
+        """Returns a set of original account_get arg names - self + new ones"""
+        func = super(AccountAnalyticDefaultAccount, self).account_get
+        return (
+            list(func.func_code.co_varnames[1:func.func_code.co_argcount]) +
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: ???

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+        date_val = kw.get('date')
+        if date_val:
+            domain += ['|', ('date_start', '<=', date_val),
+                       ('date_start', '=', False)]
+            domain += ['|', ('date_stop', '>=', date_val),
+                       ('date_stop', '=', False)]
+        return domain
+
+    def _account_get(self, **kw):
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: docstring explaining how you get to the right account

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                    index += 1
+            if index > best_index:
+                res = rec
+                best_index = index
+        return res
+
+    @api.model
+    def account_get(self, *args, **kwargs):
+        """Wrapper method to convert *args to **kwargs"""
+        kwargs.update(zip(self._account_get_domain_keys(), args))
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: don't use black magic. If the original method has some specific keys, use them.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+                    index += 1
+            if index > best_index:
+                res = rec
+                best_index = index
+        return res
+
+    @api.model
+    def account_get(self, product_id=None, partner_id=None, user_id=None,
+                    date=None, company_id=None, account_id=None):
+        """Wrapper method to convert positional args to kwargs"""
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: you don't need any wrapping:
```
@api.model
def account_get(self, product_id=None, partner_id=None, user_id=None,
                date=None, company_id=None, account_id=None):

    filters = {
        'product_id': product_id,
        'partner_id': partner_id,
        'user_id': user_id,
        'date': date,
        'company_id': company_id,
        'account_id': account_id,
    }
    domain = self._account_get_domain(**filters)
    keys = filters.keys()
    [...]
```

File: account_analytic_default_account/models/account_analytic_default_account.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        :param kw: filter keys.
+            Available filter keys are defined into `account_get_domain_keys`.
+        :return: a search domain with all required keys' leaves.
+            Each key will be searched for both False and specified value.
+            For instance: [
+                ('product_id'), '=', False), '|', [('product_id'), '=', 100)
+            ]
+        """
+        domain = []
+        account_get_domain_keys = self.account_get_domain_keys[:]
```
Comment (simahawk) [inline account_analytic_default_account/models/account_analytic_default_account.py]: you don't need this anymore as the keys are controlled by the main method signature
```
def _account_get_domain(self, **kw):
    """Build account.analytic.default domain.
    :param kw: filter keys.
        Available filter keys are defined into `account_get_domain_keys`.
    :return: a search domain with all required keys' leaves.
        Each key will be searched for both False and specified value.
        For instance: [
            ('product_id'), '=', False), '|', [('product_id'), '=', 100)
        ]
    """
    domain = []
    date_val = kw.pop('date', None)
    for key, value in kw.items():
        if value is not None:
            domain += ['|', (key, '=', value)]
        domain += [(key, '=', False)]

    if date_val:
        domain += ['|', ('date_start', '<=', date_val),
                   ('date_start', '=', False)]
        domain += ['|', ('date_stop', '>=', date_val),
                   ('date_stop', '=', False)]
    return domain
```

**Review by simahawk — CHANGES_REQUESTED** [review]

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
remember to squash before merge

**Review by leemannd — APPROVED** [review]

**Comment by fclementic2c** [comment]
cannot test it - runbot does not initate as normally :\

**Review by feketemihai — APPROVED** [review]

**Comment by feketemihai** [comment]
@pedrobaeza can you please metge this, i tested on tunbot as well..
